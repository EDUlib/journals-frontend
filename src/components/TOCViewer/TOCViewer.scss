// TODO: all colors should be variables and inheriting from site branding
.toc-border{
  width: 100%;
  background-color: #f2f8fb;
}
.toc-title{
  padding: 4px;
  font-size: 14px;
  font-weight: 600;
  font-style: normal;
  font-stretch: normal;
  line-height: 1.71;
  letter-spacing: 1px;
  color: #292b2c;
}
.toc-block{
  width: 100%;
}
.imported-toc{
  width: 100%;
  padding: 4px;
  box-sizing: border-box;
}
.responsive-toc::before {
  padding-top: 129.41%; /* about 11:8.5 standard page size */
}

