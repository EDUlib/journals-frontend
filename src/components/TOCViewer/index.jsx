import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Button, Icon } from '@edx/paragon';
import classNames from 'classnames';


import './TOCViewer.scss';


class TreeViewer extends React.Component {
  constructor(props) {
    super(props);
    this.setExpanded = this.setExpanded.bind(this);

    this.state = {
      expanded: this.props.node.id === this.props.currentPageId,
    };
  }
  componentDidUpdate(prevProps) {
    const isBeingViewed = this.props.node.id === this.props.currentPageId;
    const pageHasChanged = this.props.currentPageId !== prevProps.currentPageId;

    if (isBeingViewed && pageHasChanged && this.props.expandParent) {
      this.props.expandParent();
    }
  }

  setExpanded(event) {
    this.setState({
      expanded: event ? !this.state.expanded : true,
    });
    if (event === undefined && this.props.expandParent) {
      this.props.expandParent();
    }
  }

  render() {
    return (
      <li
        key={this.props.node.id}
        className={
          classNames({
            collapsed: this.props.node.children && !this.state.expanded,
          })
        }
      >
        <span
          className={
            classNames({
              highlight: this.props.node.id === this.props.currentPageId,
            })
          }
        >
          {
            this.props.node.children ?
              <Button
                aria-expanded={this.state.expanded}
                label={
                  <Icon
                    className={
                      classNames({
                        fa: true,
                        'fa-chevron-down': this.state.expanded,
                        'fa-chevron-right': !this.state.expanded,
                      }).split(' ')
                    }
                  />
                }
                onClick={this.setExpanded}
              /> :
              <span className="bullet">&bull;</span>
          }
          <Link to={`/${this.props.journalAboutId}/pages/${this.props.node.id}`}>{this.props.node.title}</Link>
        </span>
        {
          this.props.node.children &&
          <ul>
            <TreeList
              structure={this.props.node.children}
              journalAboutId={this.props.journalAboutId}
              currentPageId={this.props.currentPageId}
              expandParent={this.setExpanded}
            />
          </ul>
        }
      </li>
    );
  }
}

const TreeList = props => (
  props.structure.map(node => (
    <TreeViewer
      key={node.id}
      node={node}
      journalAboutId={props.journalAboutId}
      currentPageId={props.currentPageId}
      expandParent={props.expandParent}
    />
  ))
);

const TOCViewer = props => (
  // If the title is an empty string do not display one
  <div className="toc">
    <div className="toc-title">TABLE OF CONTENTS</div>
    <ul>
      <TreeList
        structure={props.journal.structure}
        journalAboutId={props.journal.id}
        currentPageId={props.currentPageId}
      />
    </ul>
  </div>
);

TreeViewer.defaultProps = {
  currentPageId: 0,
  expandParent: undefined,
};

TreeViewer.propTypes = {
  node: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    children: PropTypes.array,
  }).isRequired,
  journalAboutId: PropTypes.number.isRequired,
  currentPageId: PropTypes.number,
  expandParent: PropTypes.func,
};

TreeList.propTypes = {
  structure: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    children: PropTypes.array,
  })).isRequired,
  journalAboutId: PropTypes.number.isRequired,
};

TOCViewer.defaultProps = {
  currentPageId: 0,
};

TOCViewer.propTypes = {
  journal: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    shortDescription: PropTypes.string,
    structure: PropTypes.arrayOf(PropTypes.object),
  }).isRequired,
  currentPageId: PropTypes.number,
};

export default TOCViewer;
