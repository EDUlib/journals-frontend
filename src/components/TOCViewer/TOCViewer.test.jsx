import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import TOCViewer from '.';
import mockJournalAbout from './../JournalAboutPage/JournalAbout.mock';

const TOCViewerWrapper = props => (
  <MemoryRouter>
    <TOCViewer
      {...props}
    />
  </MemoryRouter>
);

describe('TOCViewer', () => {
  it('renders correctly on journal about page and nav panel closed', () => {
    const tree = renderer
      .create((
        <TOCViewerWrapper
          journal={mockJournalAbout}
          navPanelOpen={false}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on journal about page and nav panel opened', () => {
    const tree = renderer
      .create((
        <TOCViewerWrapper
          journal={mockJournalAbout}
          navPanelOpen
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on journal page and nav panel closed', () => {
    const tree = renderer
      .create((
        <TOCViewerWrapper
          journal={mockJournalAbout}
          currentPageId={6}
          navPanelOpen={false}
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on journal page and nav panel opened', () => {
    const tree = renderer
      .create((
        <TOCViewerWrapper
          journal={mockJournalAbout}
          currentPageId={6}
          navPanelOpen
        />
      ))
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders in expanded mode, clicks to collapse and re-renders with different current page', () => {
    const wrapper = mount((
      <TOCViewerWrapper
        journal={mockJournalAbout}
        currentPageId={8}
        navPanelOpen
      />
    ));

    let linkSelector = 'a [href="/5/pages/8"]';
    expect(wrapper.find(linkSelector).closest('span').find('span.fa-angle-down').length).toBe(1);
    wrapper.find(linkSelector).closest('span').find('button').simulate('click');
    expect(wrapper.find(linkSelector).closest('span').find('span.fa-angle-down').length).toBe(0);

    wrapper.setProps({
      journal: mockJournalAbout,
      currentPageId: 9,
      navPanelOpen: true,
    });
    linkSelector = 'a [href="/5/pages/9"]';
    expect(wrapper.find(linkSelector).closest('span').find('span.bullet').length).toBe(1);
  });
});
