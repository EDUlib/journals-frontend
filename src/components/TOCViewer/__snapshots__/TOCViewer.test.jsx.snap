// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TOCViewer renders correctly on journal about page and nav panel closed 1`] = `
<div
  className="toc"
>
  <div
    className="toc-title"
  >
    TABLE OF CONTENTS
  </div>
  <ul>
    <li
      className="collapsed"
    >
      <span
        className=""
      >
        <button
          aria-expanded={false}
          className="btn"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-angle-right"
            id="Icon1"
          />
        </button>
        <a
          href="/5/pages/6"
          onClick={[Function]}
          tabIndex={-1}
        >
          Title of journal page 1
        </a>
      </span>
      <ul>
        <li
          className="collapsed"
        >
          <span
            className=""
          >
            <button
              aria-expanded={false}
              className="btn"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-right"
                id="Icon1"
              />
            </button>
            <a
              href="/5/pages/8"
              onClick={[Function]}
              tabIndex={-1}
            >
              Title of journal page 1 child
            </a>
          </span>
          <ul>
            <li
              className=""
            >
              <span
                className=""
              >
                <span
                  className="bullet"
                >
                  <div>
                    •
                  </div>
                </span>
                <a
                  href="/5/pages/9"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Title of journal page 1 grand child
                </a>
              </span>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className=""
    >
      <span
        className=""
      >
        <span
          className="bullet"
        >
          <div>
            •
          </div>
        </span>
        <a
          href="/5/pages/7"
          onClick={[Function]}
          tabIndex={-1}
        >
          Title of journal page 2
        </a>
      </span>
    </li>
  </ul>
</div>
`;

exports[`TOCViewer renders correctly on journal about page and nav panel opened 1`] = `
<div
  className="toc"
>
  <div
    className="toc-title"
  >
    TABLE OF CONTENTS
  </div>
  <ul>
    <li
      className="collapsed"
    >
      <span
        className=""
      >
        <button
          aria-expanded={false}
          className="btn"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-angle-right"
            id="Icon1"
          />
        </button>
        <a
          href="/5/pages/6"
          onClick={[Function]}
          tabIndex={0}
        >
          Title of journal page 1
        </a>
      </span>
      <ul>
        <li
          className="collapsed"
        >
          <span
            className=""
          >
            <button
              aria-expanded={false}
              className="btn"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-right"
                id="Icon1"
              />
            </button>
            <a
              href="/5/pages/8"
              onClick={[Function]}
              tabIndex={-1}
            >
              Title of journal page 1 child
            </a>
          </span>
          <ul>
            <li
              className=""
            >
              <span
                className=""
              >
                <span
                  className="bullet"
                >
                  <div>
                    •
                  </div>
                </span>
                <a
                  href="/5/pages/9"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Title of journal page 1 grand child
                </a>
              </span>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className=""
    >
      <span
        className=""
      >
        <span
          className="bullet"
        >
          <div>
            •
          </div>
        </span>
        <a
          href="/5/pages/7"
          onClick={[Function]}
          tabIndex={0}
        >
          Title of journal page 2
        </a>
      </span>
    </li>
  </ul>
</div>
`;

exports[`TOCViewer renders correctly on journal page and nav panel closed 1`] = `
<div
  className="toc"
>
  <div
    className="toc-title"
  >
    TABLE OF CONTENTS
  </div>
  <ul>
    <li
      className=""
    >
      <span
        className="highlight"
      >
        <button
          aria-expanded={true}
          className="btn"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-angle-down"
            id="Icon1"
          />
        </button>
        <a
          href="/5/pages/6"
          onClick={[Function]}
          tabIndex={-1}
        >
          Title of journal page 1
        </a>
      </span>
      <ul>
        <li
          className="collapsed"
        >
          <span
            className=""
          >
            <button
              aria-expanded={false}
              className="btn"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={-1}
              type="button"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-right"
                id="Icon1"
              />
            </button>
            <a
              href="/5/pages/8"
              onClick={[Function]}
              tabIndex={-1}
            >
              Title of journal page 1 child
            </a>
          </span>
          <ul>
            <li
              className=""
            >
              <span
                className=""
              >
                <span
                  className="bullet"
                >
                  <div>
                    •
                  </div>
                </span>
                <a
                  href="/5/pages/9"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Title of journal page 1 grand child
                </a>
              </span>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className=""
    >
      <span
        className=""
      >
        <span
          className="bullet"
        >
          <div>
            •
          </div>
        </span>
        <a
          href="/5/pages/7"
          onClick={[Function]}
          tabIndex={-1}
        >
          Title of journal page 2
        </a>
      </span>
    </li>
  </ul>
</div>
`;

exports[`TOCViewer renders correctly on journal page and nav panel opened 1`] = `
<div
  className="toc"
>
  <div
    className="toc-title"
  >
    TABLE OF CONTENTS
  </div>
  <ul>
    <li
      className=""
    >
      <span
        className="highlight"
      >
        <button
          aria-expanded={true}
          className="btn"
          onBlur={[Function]}
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            aria-hidden={true}
            className="fa fa-angle-down"
            id="Icon1"
          />
        </button>
        <a
          href="/5/pages/6"
          onClick={[Function]}
          tabIndex={0}
        >
          Title of journal page 1
        </a>
      </span>
      <ul>
        <li
          className="collapsed"
        >
          <span
            className=""
          >
            <button
              aria-expanded={false}
              className="btn"
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-right"
                id="Icon1"
              />
            </button>
            <a
              href="/5/pages/8"
              onClick={[Function]}
              tabIndex={0}
            >
              Title of journal page 1 child
            </a>
          </span>
          <ul>
            <li
              className=""
            >
              <span
                className=""
              >
                <span
                  className="bullet"
                >
                  <div>
                    •
                  </div>
                </span>
                <a
                  href="/5/pages/9"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Title of journal page 1 grand child
                </a>
              </span>
            </li>
          </ul>
        </li>
      </ul>
    </li>
    <li
      className=""
    >
      <span
        className=""
      >
        <span
          className="bullet"
        >
          <div>
            •
          </div>
        </span>
        <a
          href="/5/pages/7"
          onClick={[Function]}
          tabIndex={0}
        >
          Title of journal page 2
        </a>
      </span>
    </li>
  </ul>
</div>
`;
